substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -1417,9 +1417,70 @@ where
 		// transactions with those hashes from the retracted blocks.
 		let mut pruned_log = HashSet::<ExtrinsicHash<ChainApi>>::new();
 
+		// Create a map from (retracted blocks + common block)'s extrinsics to their `provides`
+		// tags.
+		let inactive_views_xts_to_tags = Arc::new(
+			future::join_all(
+				tree_route
+					.retracted()
+					.iter()
+					.chain(std::iter::once(tree_route.common_block()))
+					.map(|hn| {
+						let api = api.clone();
+						async move {
+							// Fetch extrinsics.
+							let view = self.view_store.get_view_at(hn.hash, true);
+							let xts = api
+								.block_body(hn.hash)
+								.await
+								.unwrap_or_else(|e| {
+									log::warn!(target: LOG_TARGET, "Block txs: error request: {}", e);
+									None
+								})
+								.unwrap_or_default();
+
+							let xts_hashes = xts
+								.iter()
+								.filter_map(|xts| {
+									view.as_ref().map(|(inner, _)| inner.pool.hash_of(xts))
```

Comment:
tuple `(hash,provide_tag)` could be returned here to avoid zip and to have single iteration?

Reply:
If the view is none we can skip entire logic earlier - no need to filter_map on view being Some here.

Reply:
> tuple (hash,provide_tag) could be returned here to avoid zip and to have single iteration?

Sorry I re-thought this, zipping probably makes sense here (as tags function accepts array).